### PR TITLE
use lock-free freelist

### PR DESCRIPTION
**Pull request by MartinNowak**

### CODE REVIEW

**Comment by braddr** [comment]
I'm believe that these are subject to classic ABA problems.  http://en.wikipedia.org/wiki/ABA_problem

**Comment by andralex** [comment]
@MartinNowak are you subject to ABA? I probably won't have time to look closer into this.

**Comment by Safety0ff** [comment]
Yup, it's subject to ABA.
Lockfree LIFO stack on x86 recquires DWCAS which isn't always available on x86_64 (early amd64 CPUs.)
On x86_64 you'd either have to switch to manual memory management + abuse upper 16 bits of pointers (to store a counter) or use FIFO instead.
There's also no guarantee that the lockfree solution would beat a simple spinlock with sleep/yield:
http://blog.memsql.com/common-pitfalls-in-writing-lock-free-algorithms/
That said, switching to manual memory management for the nodes is probably a good idea as long as we're using a freelist.

**Comment by MartinNowak** [comment]
> I'm believe that these are subject to classic ABA problems. http://en.wikipedia.org/wiki/ABA_problem

Yeah, you're right.
Here is a related pull to fix `cas` for structs.
https://github.com/D-Programming-Language/druntime/pull/1100

**Comment by MartinNowak** [comment]
I went with a spinlock instead of DWCAS as that's more portable, not much speed differences though.
For the contended case this now calls `Thread.yield` which helps a little to avoid burning cycles.
I also destroy the value before putting it on the freelist to not pin any GC memory.

**Comment by MartinNowak** [comment]
> That said, switching to manual memory management for the nodes is probably a good idea as long as we're using a freelist.

No, it's not because we need to register every Node as GC root.

**Comment by MartinNowak** [comment]
> or use FIFO instead

Any reference for a good FIFO implementation? I'm actually collecting lock-free containers. Want to contribute an implementation or two?
https://github.com/MartinNowak/lock-free

**Comment by Safety0ff** [comment]
> not much speed differences though

Not much difference vs mutex or not much difference vs the lockfree (with/without ABA prevention)?

> No, it's not because we need to register every Node as GC root.

Oops, you're right.

> Any reference for a good FIFO implementation? I'm actually collecting lock-free containers. Want to contribute an implementation or two?
https://github.com/MartinNowak/lock-free

These come to mind:
http://www.1024cores.net/home/lock-free-algorithms/queues/non-intrusive-mpsc-node-based-queue
http://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue

**Comment by MartinNowak** [comment]
> Not much difference vs mutex or not much difference vs the lockfree (with/without ABA prevention)?

Not much difference of DWCAS and spinlock.

**Comment by MartinNowak** [comment]
Anyone?

File: std/concurrency.d [hunk 1]
```diff
@@ -2317,73 +2317,46 @@ private
...
-
-        __gshared Node* sm_freeNodes = null;
-        __gshared Mutex sm_freeLock  = null;
-
-
-        shared static this()
+        static shared struct SpinLock
         {
-            sm_freeLock = new Mutex;
+            void lock() { while (!cas(&locked, false, true)) { Thread.yield(); } }
```
Comment (andralex) [inline std/concurrency.d]: I think calling yield with that frequency is not the best way to go.

**Comment by andralex** [comment]
will merge now, subject to future possible improvements

**Comment by andralex** [comment]
thanks!

**Pull request merged**